timeseries/src/main/scala/com/criteo/cuttle/timeseries/TimeSeriesScheduler.scala

```diff
@@ -420,41 +420,48 @@ case class TimeSeriesScheduler(logger: Logger) extends Scheduler[TimeSeries] wit
                                       xa: XA)(implicit user: Auth.User): IO[Either[String, Unit]] = {
 
+    logger.debug(s"Requesting a backfill of ${jobs.map(_.id)} between $start and $end")
 
-    val backfillErrors: List[String] = atomic { implicit txn =>
-      val errors = mutable.ArrayBuffer.empty[String]
-      val backfills = createBackfills(name, description, jobs, _state(), start, end, priority)
+    val result = atomic { implicit txn =>
+      val currentJobStates = _state()
+      val backfillErrors = mutable.ArrayBuffer.empty[String]
+      val backfills = createBackfills(name, description, jobs, currentJobStates, start, end, priority)
 
       val validBackfills: List[Backfill] = backfills.flatMap { newBackfill =>
-        validateBackfill(newBackfill, _state()).toList match {
+        validateBackfill(newBackfill, currentJobStates).toList match {
           case Nil =>  Some(newBackfill)
           case (validationErrors) =>
-            errors += validationErrors.mkString("\n")
+            backfillErrors += validationErrors.mkString("\n")
             None
         }
       }
 
-      if (errors.isEmpty) {
-        val executionsToCancel = listExecutionsToCancelGivenBackfills(validBackfills, _state(), runningExecutions.keys)
-        executionsToCancel.foreach { e =>
-          logger.info(s"Cancelling running execution ${e.id} for job ${e.job.id}")
-          e.cancel()
+      if (backfillErrors.isEmpty) {
+        val (concurrentExecutions, concurrentExecutionsPromise) = {
+          val executionsToCancel = listExecutionsToCancelGivenBackfills(validBackfills, currentJobStates, runningExecutions.keys)
+          (executionsToCancel, Future.sequence(executionsToCancel.flatMap(e => runningExecutions.get(e))))
         }
-        Future.sequence(executionsToCancel.map(runningExecutions.apply))
-          .andThen { case _ => validBackfills.foreach { newBackfill =>
-            updateBackfillState(newBackfill)
-            Database.createBackfill(newBackfill).transact(xa)
-          } }
+
+        Right((validBackfills, concurrentExecutions, concurrentExecutionsPromise))
+      }
+      else {
+        Left(backfillErrors.toList)
       }
-      errors.toList
     }
 
-
-
-    if (backfillErrors.nonEmpty)
-      IO.pure(Left(backfillErrors.mkString("\n")))
-    else
-      IO.pure(Right(Unit))
+    result match {
+      case Left(errors) => IO.pure(Left(errors.mkString("\n")))
+      case Right((backfills, concurrentExecutions, concurrentExecutionsPromise)) =>
+        concurrentExecutions.foreach { e =>
+          logger.info (s"Cancelling running execution ${e.id} for job ${e.job.id}")
+          e.cancel()
+        }
+        concurrentExecutionsPromise.andThen { case _ => backfills.foreach { newBackfill =>
+          updateBackfillState(newBackfill)
+          Database.createBackfill(newBackfill).transact(xa).unsafeRunSync
```

Comment:
Generally we don't use unsafeRunSync everywhere, try to construct an IO flow with `map/flatMap` or `IO.fromFuture` and return this constructed IO. Btw you can catch some exceptions with database interactions.

Reply:
We can check this together the next week.
Aside from this cancelling concurrent executions when requesting a backfill should not be done imo (if there is such a need then the user should do it). I will remove the cancellation part.